mkosi/gentoo.py

```diff
@@ -0,0 +1,508 @@
+        # apply whatever we put in gentoo_mkosi_conf to runs invokation of emerge
+        with open(package_env / "mkosi", "a") as f:
+            f.write("*/*    mkosi.conf\n")
+
+        uninstall_ignore = 'UNINSTALL_IGNORE="${UNINSTALL_IGNORE} ' + " ".join(self.UNINSTALL_IGNORE) + '"'
+        use = 'USE="${USE} ' + " ".join(self.portage_use_flags) + '"'
+        with open(gentoo_mkosi_conf, "a") as f:
+            f.write(
+                dedent(
+                    f"""\
+                    # MKOSI: these were used during image creation...
+                    # and some more! see under package.*/
+                    #
+                    # usrmerge (see all under profile/)
+                    {uninstall_ignore}
+                    {use}
+                    """
+                )
+            )
+
+    def invoke_emerge(
+        self,
+        args: CommandLineArguments,
+        root: str,
+        inside_stage3: bool = True,
+        pkgs: Set[str] = set(),
+        actions: Set[str] = set(),
+        opts: Set[str] = set(),
+    ) -> None:
+        if not inside_stage3:
+            from _emerge.main import emerge_main  # type: ignore
+
+            os.environ["FEATURES"] = " ".join(self.portage_features)
+            os.environ["BOOTSTRAP_USE"] = " ".join(self.portage_use_flags)
+            os.environ["USE"] = " ".join(self.portage_use_flags)
+            os.environ["GRUB_PLATFORMS"] = " ".join(self.grub_platforms)
+            os.environ["EGIT_CLONE_TYPE"] = "shallow"
+
+            PREFIX_OPTS: Set[str] = set()
+            if "--sync" not in actions:
+                PREFIX_OPTS = {
+                    "--config-root=" + root,
+                    "--root=" + root,
+                    "--sysroot=" + root,
+                }
+
+            emerge_main(pkgs.union(opts, PREFIX_OPTS, self.emerge_default_opts, actions))
+        else:
+            cmdline = ["/usr/bin/emerge"]
+
+            emerge_env = {
+                "FEATURES": " ".join(self.portage_features),
+                "BOOTSTRAP_USE": " ".join(self.portage_use_flags),
+                "USE": " ".join(self.portage_use_flags),
+                "GRUB_PLATFORMS": " ".join(self.grub_platforms),
+                "EGIT_CLONE_TYPE": "shallow",
+            }
+
+            cmdline.extend(pkgs)
+            cmdline.extend(self.emerge_default_opts)
+            cmdline.extend(opts)
+            cmdline.extend(actions)
+
+            MkosiPrinter.print_step("Invoking emerg(1) inside stage3")
+            run_workspace_command(
+                args,
+                root,
+                cmdline,
+                network=True,
+                env=emerge_env,
+                nspawn_params=[
+                    "--capability=CAP_SYS_ADMIN,CAP_MKNOD",
+                    "--bind=" + self.portage_cfg["PORTDIR"],
+                    "--bind=" + self.portage_cfg["DISTDIR"],
+                    "--bind=" + self.portage_cfg["PKGDIR"],
```

Comment:
Those lines repeat, shouldn't they be abstracted to some helper?

Reply:
added to globals, let me know.